### PR TITLE
Fix dead code warning when inline const is used in pattern

### LINKED ISSUE #78171: dead_code lint triggers even though function is used in const pattern
This should compile without warnings ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0fc7c5eb8cb3a350c4bce0c1ab8b5676)):

```rust
#![allow(incomplete_features)]
#![feature(inline_const)]

const fn one() -> i32 {
    1
}

fn main() {
    match 1 {
        const { one() } => {}
        _ => {}
    }
}
```

Yet it warns:

```
   Compiling playground v0.0.1 (/playground)
warning: function is never used: `one`
 --> src/main.rs:4:10
  |
4 | const fn one() -> i32 {
  |          ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.80s
     Running `target/debug/playground`
```

### PR DESCRIPTION
**Pull request by nbdd0121**
Fixes #78171

### CODE REVIEW

**Comment by petrochenkov** [comment]
I'd never guess that `const { ... }` patterns use `PatKind::Lit` internally.
@bors r+

**Comment by nbdd0121** [comment]
Actually this approach doesn't catch `const { foo() }..2` case. Might be better to save `in_pat` and set it to false when visiting a `AnonConst` and restore it after visiting.

**Comment by nbdd0121** [comment]
Added a test case for using inline const in range pattern

**Comment by nbdd0121** [comment]
> I'd never guess that `const { ... }` patterns use `PatKind::Lit` internally.

It's surprising to me as well, but I guess that's probably the easiest way to implement it since it'll need to have a `P<Expr>` anyway.

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -395,8 +396,14 @@ impl<'tcx> Visitor<'tcx> for MarkSymbolVisitor<'tcx> {
     }
 
     fn visit_anon_const(&mut self, c: &'tcx hir::AnonConst) {
+        // When inline const blocks are used in pattern position, paths
+        // refereneced by it should be considered as used.
```
Comment (klensy) [inline compiler/rustc_passes/src/dead.rs:400]:
```suggestion
        // referenced by it should be considered as used.
```

**Comment by petrochenkov** [comment]
>Might be better to save in_pat and set it to false when visiting a AnonConst and restore it after visiting.

Yeah, it's a better approach, it should also fix cases like
```rust
const fn one() -> i32 {
    1
}

struct S<const C: i32>;

fn main() {
    match S::<1> {
        S::<{one()}> => {}
    }
}
```

Could you actually add this case too?
And merge all the cases into a single test file, no need to run rustc multiple times to test all of this.

@bors r-

**Comment by nbdd0121** [comment]
Added and merged

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**